x-pack/plugin/esql/src/internalClusterTest/java/org/elasticsearch/xpack/esql/action/LookupJoinTypesIT.java

```diff
@@ -0,0 +1,463 @@
+ * in the lookup index. This can be confusing, but it is important to remember that the field names are not the same as the types.
+ * For example, if we are testing the pairs (double, double), (double, float), (float, double) and (float, float),
+ * we will create the following indexes:
+ * <dl>
+ *     <dt>index_double_double</dt>
+ *     <dd>Index containing a single document with a field of type 'double' like: <pre>
+ *         {
+ *             "field_double": 1.0,  // this is mapped as type 'double'
+ *             "other": "value"
+ *         }
+ *     </pre></dd>
+ *     <dt>index_double_float</dt>
+ *     <dd>Index containing a single document with a field of type 'float' like: <pre>
+ *         {
+ *             "field_double": 1.0,  // this is mapped as type 'float' (a float with the name of the main index field)
+ *             "other": "value"
+ *         }
+ *     </pre></dd>
+ *     <dt>index_float_double</dt>
+ *     <dd>Index containing a single document with a field of type 'double' like: <pre>
+ *         {
+ *             "field_float": 1.0,  // this is mapped as type 'double' (a double with the name of the main index field)
+ *             "other": "value"
+ *         }
+ *     </pre></dd>
+ *     <dt>index_float_float</dt>
+ *     <dd>Index containing single document with a field of type 'float' like: <pre>
+ *         {
+ *             "field_float": 1.0,  // this is mapped as type 'float'
+ *             "other": "value"
+ *         }
+ *     </pre></dd>
+ *     <dt>index</dt>
+ *     <dd>Index containing document like: <pre>
+ *         {
+ *             "field_double": 1.0,  // this is mapped as type 'double'
+ *             "field_float": 1.0    // this is mapped as type 'float'
+ *         }
+ *     </pre></dd>
+ * </dl>
+ * Note that the lookup indexes have fields with a name that matches the type in the main index, and not the type actually used in the
+ * lookup index. Instead, the mapped type should be the type of the right-hand side of the pair being tested.
+ * Then we can run queries like:
+ * <pre>
+ *     FROM index | LOOKUP JOIN index_double_float ON field_double | KEEP other
+ * </pre>
+ * And assert that the result exists and is equal to "value".
+ */
+@ClusterScope(scope = SUITE, numClientNodes = 1, numDataNodes = 1)
+public class LookupJoinTypesIT extends ESIntegTestCase {
+    protected Collection<Class<? extends Plugin>> nodePlugins() {
+        return List.of(EsqlPlugin.class, MapperExtrasPlugin.class);
+    }
+
+    private static final Map<String, TestConfigs> testConfigurations = new HashMap<>();
+    static {
+        // Initialize the test configurations for string tests
+        {
+            TestConfigs configs = testConfigurations.computeIfAbsent("strings", TestConfigs::new);
+            configs.addPasses(KEYWORD, KEYWORD);
+            configs.addPasses(TEXT, KEYWORD);
```

Comment:
Is it possible to have a `TEXT` field without a `.keyword` subfield? I think we fail in this situation. Could you maybe double check and add a test if that's possible?

Reply:
We always pass with TEXT on the left because we will read from source for the main index if no KEYWORD subfield exists. Likewise we always fail with TEXT on the right, even if a KEYWORD subfield exists because we have not coded a special case for that (yet).